cmd/dep/composite_analyzer.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2016 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"github.com/golang/dep"
+	"github.com/golang/dep/internal/gps"
+	"github.com/pkg/errors"
+)
+
+// compositeAnalyzer overlays configuration from multiple analyzers
+type compositeAnalyzer struct {
+	// Analyzers is the set of analyzers to apply, last one wins any conflicts
+	Analyzers []rootProjectAnalyzer
+}
+
+func (a compositeAnalyzer) DeriveRootManifestAndLock(path string, n gps.ProjectRoot) (*dep.Manifest, *dep.Lock, error) {
+	var rootM *dep.Manifest
+	var rootL *dep.Lock
+
+	for _, a := range a.Analyzers {
+		m, l, err := a.DeriveRootManifestAndLock(path, n)
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, "%T.DeriveRootManifestAndLock", a)
+		}
+
+		if rootM == nil && rootL == nil {
+			rootM = m
+			rootL = l
+		} else {
+			// Overlay the changes from the analyzer on-top of the previous analyzer's work
+			if m != nil {
+				for pkg, prj := range m.Dependencies {
+					rootM.Dependencies[pkg] = prj
+				}
+				for pkg, prj := range m.Ovr {
+					rootM.Ovr[pkg] = prj
+				}
+				for _, pkg := range m.Required {
+					if !contains(rootM.Required, pkg) {
+						rootM.Required = append(rootM.Required, pkg)
+					}
+				}
+				for _, pkg := range m.Ignored {
+					if !contains(rootM.Ignored, pkg) {
+						rootM.Ignored = append(rootM.Ignored, pkg)
+					}
+				}
+			}
+
+			if l != nil {
+				for _, lp := range l.P {
+					for i, existingLP := range rootL.P {
+						if lp.Ident().ProjectRoot == existingLP.Ident().ProjectRoot {
+							rootL.P[i] = lp
+						}
+					}
+				}
```

Comment:
@carolynvs a little problem here. Overlaying of lock isn't happening in some cases.

Let's take the example of integration test init/godep/case1/.
In godep/case1/testcase.json, although `gopath-initial` contains `deptest`, since deptest is a transitive dependency, the default (first analyzer) doesn't adds it to the manifest or lock it generates. And since there's no `deptestdos` in `gopath-initial`, the first analyzer generates empty (not nil) manifest and lock structs.
Now in the above code, when overlay is attempted for lock with godep analyzer generated manifest and lock,
```
for i, existingLP := range rootL.P {
```
is skipped because `rootL` is empty from the first analyzer run. And no lock overlay happens.

Makes sense?

The final lock in init/godep/case1/ is incorrect at the moment. Commit hash in final lock is different from the one in Godep.json.